### PR TITLE
Convnext flax

### PR DESCRIPTION
**Pull request by Shubhamai**
# Flax Implementation of `facebook/convnext-tiny-224`

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?

## Who can review?

Flax: @sanchit-gandhi

## TODO

Last Updated : 10 Feb, 2023

- [x] Fixing tests failed in `ci/circleci: tests_flax` actions.   
- [ ] Uploading [Shubhamai/convnext-tiny-224](https://huggingface.co/Shubhamai/convnext-tiny-224) flax weights to [facebook/convnext-tiny-224](https://huggingface.co/facebook/convnext-tiny-224).
- [x] Depends on merge of [_Improve flax test_attention_outputs_ PR](https://github.com/huggingface/transformers/pull/21486) to pass (or technically skip) `test_attention_outputs` test.

### CODE REVIEW

**Comment by Shubhamai** [comment]
@sanchit-gandhi this PR is also ready for your review in case it was missed. And again, thank you so much for taking the time to review the PR.

**Comment by Shubhamai** [comment]
@sanchit-gandhi Reminder incase my previous message got missed! Also the https://github.com/huggingface/transformers/pull/21472 ( previous reviews implemented ) and https://github.com/huggingface/transformers/pull/21867 PR are awaiting review. Thanks!

File: src/transformers/models/convnext/configuration_convnext.py [hunk 1]
```diff
@@ -64,6 +64,8 @@ class ConvNextConfig(PretrainedConfig):
             The initial value for the layer scale.
         drop_path_rate (`float`, *optional*, defaults to 0.0):
             The drop rate for stochastic depth.
+        image_size (`int`, *optional*, defaults to 224):
```
Comment (sanchit-gandhi) [inline src/transformers/models/convnext/configuration_convnext.py:67]: Think we can just remove this from the docstring (see https://github.com/huggingface/transformers/pull/21472#discussion_r1118985964)

File: src/transformers/models/convnext/modeling_flax_convnext.py [hunk 2]
```diff
@@ -0,0 +1,627 @@
...
+    See discussion: https://github.com/tensorflow/tpu/issues/494#issuecomment-532968956 ... I've opted for changing the
+    layer and argument names to 'drop path' rather than mix DropConnect as a layer name and use 'survival rate' as the
+    argument.
+    """
+
+    drop_prob: Optional[float] = None
+    dtype: jnp.dtype = jnp.float32
+
+    @nn.compact
+    def __call__(self, hidden_states: jnp.ndarray, train: bool = False) -> jnp.ndarray:
```
Comment (sanchit-gandhi) [inline src/transformers/models/convnext/modeling_flax_convnext.py:57]: We can also swap `train` for `deterministic` (see https://github.com/huggingface/transformers/pull/21472#discussion_r1118999847)

File: src/transformers/models/convnext/modeling_flax_convnext.py [hunk 3]
```diff
@@ -0,0 +1,627 @@
...
+        if self.drop_prob == 0.0 or not train:
+            return hidden_states
+        keep_prob = 1 - self.drop_prob
+        shape = (hidden_states.shape[0],) + (1,) * (
+            hidden_states.ndim - 1
+        )  # work with diff dim tensors, not just 2D ConvNets
+        random_tensor = keep_prob + random.uniform(random.PRNGKey(0), shape, dtype=self.dtype)
```
Comment (sanchit-gandhi) [inline src/transformers/models/convnext/modeling_flax_convnext.py:64]: We probably need to pass a PRNG key here (rather than always starting from seed 0), see https://github.com/huggingface/transformers/blob/90247d3e011fef58e2b4e95f70ae2c1aacaf61db/src/transformers/models/bert/modeling_flax_bert.py#L364-L366 and https://github.com/huggingface/transformers/pull/21023#pullrequestreview-1247514135

File: src/transformers/models/convnext/modeling_flax_convnext.py [hunk 4]
```diff
@@ -0,0 +1,627 @@
...
+            kernel_size=(self.config.patch_size, self.config.patch_size),
+            strides=(self.config.patch_size, self.config.patch_size),
+            dtype=self.dtype,
+            kernel_init=jax.nn.initializers.normal(self.config.initializer_range),
+        )
+        self.layernorm = nn.LayerNorm(epsilon=1e-6, dtype=self.dtype)
+
+    def __call__(self, pixel_values: jnp.ndarray) -> jnp.ndarray:
+        num_channels = pixel_values.shape[-1]
+        if num_channels != self.config.num_channels:
```
Comment (sanchit-gandhi) [inline src/transformers/models/convnext/modeling_flax_convnext.py:98]: Same here regarding the config and `num_channels`

File: src/transformers/models/convnext/modeling_flax_convnext.py [hunk 5]
```diff
@@ -0,0 +1,627 @@
...
+        return embeddings
+
+
+class FlaxConvNextLayer(nn.Module):
+    """This corresponds to the `Block` class in the original implementation.
+
+    There are two equivalent implementations: [DwConv, LayerNorm (channels_first), Conv, GELU,1x1 Conv]; all in (N, C,
+    H, W) (2) [DwConv, Permute to (N, H, W, C), LayerNorm (channels_last), Linear, GELU, Linear]; Permute back
+
+    The authors used (2) as they find it slightly faster in PyTorch. This Flax implementation is based on (1)
```
Comment (sanchit-gandhi) [inline src/transformers/models/convnext/modeling_flax_convnext.py:113]: Cool!

File: src/transformers/models/convnext/modeling_flax_convnext.py [hunk 6]
```diff
@@ -0,0 +1,627 @@
...
+        self.layernorm = nn.LayerNorm(epsilon=1e-6, dtype=self.dtype)
+        self.pwconv1 = nn.Dense(4 * self.dim, dtype=self.dtype)  # pointwise/1x1 convs, implemented with linear layers
+        self.act = ACT2FN[self.config.hidden_act]
+        self.pwconv2 = nn.Dense(self.dim, dtype=self.dtype)
+
+        layer_scale_init_value = self.config.layer_scale_init_value if self.config.layer_scale_init_value > 0 else 1
+        layer_scale_parameter_init = jax.nn.initializers.constant(layer_scale_init_value)
+        self.layer_scale_parameter = self.param("layer_scale_parameter", layer_scale_parameter_init, (self.dim))
+
+        self.drop_path_func = FlaxConvNextDropPath(self.drop_path, dtype=self.dtype) if self.drop_path > 0.0 else None
```
Comment (sanchit-gandhi) [inline src/transformers/models/convnext/modeling_flax_convnext.py:145]: We've defined the Identity function above so can use this here if self.drop_path = 0.0 (as we do in the PyTorch code)

File: src/transformers/models/convnext/modeling_flax_convnext.py [hunk 7]
```diff
@@ -0,0 +1,627 @@
...
+
+    config_class = ConvNextConfig
+    base_model_prefix = "convnext"
+    main_input_name = "pixel_values"
+    module_class: nn.Module = None
+
+    def __init__(
+        self,
+        config: ConvNextConfig,
+        input_shape=None,
```
Comment (sanchit-gandhi) [inline src/transformers/models/convnext/modeling_flax_convnext.py:344]: See comment on `input_shape`: https://github.com/huggingface/transformers/pull/21472#discussion_r1118991044

File: src/transformers/models/convnext/modeling_flax_convnext.py [hunk 8]
```diff
@@ -0,0 +1,627 @@
...
+        if not return_dict:
+            return tuple(v for v in [hidden_states, all_hidden_states] if v is not None)
+
+        return FlaxBaseModelOutputWithNoAttention(
+            last_hidden_state=hidden_states,
+            hidden_states=all_hidden_states,
+        )
+
+
+class FlaxConvNextPreTrainedModel(FlaxPreTrainedModel):
```
Comment (sanchit-gandhi) [inline src/transformers/models/convnext/modeling_flax_convnext.py:330]: Can this now be `# Copied from ...` one of our Flax ResNet models?

File: src/transformers/models/convnext/modeling_flax_convnext.py [hunk 9]
```diff
@@ -0,0 +1,627 @@
...
+        pixel_values: jnp.ndarray = None,
+        output_hidden_states: Optional[bool] = None,
+        return_dict: Optional[bool] = None,
+    ) -> Union[Tuple, FlaxBaseModelOutputWithPoolingAndNoAttention]:
+        output_hidden_states = (
+            output_hidden_states if output_hidden_states is not None else self.config.output_hidden_states
+        )
+        return_dict = return_dict if return_dict is not None else self.config.use_return_dict
+
+        if pixel_values is None:
```
Comment (sanchit-gandhi) [inline src/transformers/models/convnext/modeling_flax_convnext.py:479]: This won't play very nicely with JIT since we have control flow on our input (see https://jax.readthedocs.io/en/latest/notebooks/Common_Gotchas_in_JAX.html#control-flow)

Think we can safely assume here that `pixel_values` will not be None (otherwise the next line `self.embeddings(pixel_values)` will complain anyway that `pixel_values` is None), and so can remove this check.

File: src/transformers/models/convnext/modeling_flax_convnext.py [hunk 10]
```diff
@@ -0,0 +1,627 @@
...
+    ```python
+    >>> from transformers import AutoImageProcessor, FlaxConvNextModel
+    >>> from PIL import Image
+    >>> import requests
+
+    >>> url = "http://images.cocodataset.org/val2017/000000039769.jpg"
+    >>> image = Image.open(requests.get(url, stream=True).raw)
+
+    >>> image_processor = AutoImageProcessor.from_pretrained("facebook/convnext-tiny-224")
+    >>> model = FlaxConvNextModel.from_pretrained("facebook/convnext-tiny-224")
```
Comment (sanchit-gandhi) [inline src/transformers/models/convnext/modeling_flax_convnext.py:529]: Cool! Have we uploaded the Flax weights already?

File: tests/models/convnext/test_modeling_flax_convnext.py [hunk 11]
```diff
@@ -0,0 +1,262 @@
...
+
+    from transformers import FlaxConvNextForImageClassification, FlaxConvNextModel
+
+if is_vision_available():
+    from PIL import Image
+
+    from transformers import AutoFeatureExtractor
+
+
+class FlaxConvNextModelTester(unittest.TestCase):
```
Comment (sanchit-gandhi) [inline tests/models/convnext/test_modeling_flax_convnext.py:38]: Would be awesome to update the tester according to the Flax ResNet model tester :)

File: tests/models/convnext/test_modeling_flax_convnext.py [hunk 12]
```diff
@@ -0,0 +1,262 @@
...
+
+@require_flax
+class FlaxConvNextModelIntegrationTest(unittest.TestCase):
+    @cached_property
+    def default_feature_extractor(self):
+        return AutoFeatureExtractor.from_pretrained("facebook/convnext-tiny-224") if is_vision_available() else None
+
+    @slow
+    def test_inference_image_classification_head(self):
+        model = FlaxConvNextForImageClassification.from_pretrained("Shubhamai/convnext-tiny-224")
```
Comment (sanchit-gandhi) [inline tests/models/convnext/test_modeling_flax_convnext.py:248]: Feel free to open a PR for the Flax weights on the Facebook org!

**Review by sanchit-gandhi — COMMENTED** [review]
Looks great already @Shubhamai! Think we just need to propagate some of the changes from Flax ResNet / RegNet into this PR, then we're good to go 🚀

File: src/transformers/models/convnext/modeling_flax_convnext.py (hunk 8 above)
Comment (Shubhamai) [reply src/transformers/models/convnext/modeling_flax_convnext.py:330]: No, because we don't have the batch norm layers, we cannot input `batch_stats` here

```python
self.module.apply(
            {"params": params or self.params}
```

File: src/transformers/models/convnext/modeling_flax_convnext.py (hunk 10 above)
Comment (Shubhamai) [reply src/transformers/models/convnext/modeling_flax_convnext.py:529]: Just made the PR https://huggingface.co/facebook/convnext-tiny-224/discussions/3

File: tests/models/convnext/test_modeling_flax_convnext.py (hunk 12 above)
Comment (Shubhamai) [reply tests/models/convnext/test_modeling_flax_convnext.py]: PR https://huggingface.co/facebook/convnext-tiny-224/discussions/3

File: src/transformers/models/convnext/modeling_flax_convnext.py (hunk 4 above)
Comment (Shubhamai) [reply src/transformers/models/convnext/modeling_flax_convnext.py]: I assume you mean removing the if statement and `num_channels` in the config file. If yes, then the changed should be made in the recent commit

File: src/transformers/models/convnext/modeling_flax_convnext.py (hunk 3 above)
Comment (Shubhamai) [reply src/transformers/models/convnext/modeling_flax_convnext.py]: Replaced `random.PRNGKey(0)` with `self.make_rng("dropout") `

File: src/transformers/models/convnext/modeling_flax_convnext.py (hunk 6 above)
Comment (Shubhamai) [reply src/transformers/models/convnext/modeling_flax_convnext.py]: Agree, should be fixed in recent commit

File: src/transformers/models/convnext/configuration_convnext.py [hunk 13]
```diff
@@ -90,7 +90,6 @@ class ConvNextConfig(PretrainedConfig):
 
     def __init__(
         self,
-        num_channels=3,
```
Comment (sanchit-gandhi) [inline src/transformers/models/convnext/configuration_convnext.py:93]: We can't remove `config.num_channels` since this will break the PyTorch modelling code! https://github.com/huggingface/transformers/blob/40082d598b756c4f1fb048c571ed95c31f05a69c/src/transformers/models/convnext/modeling_convnext.py#LL137C2-L137C2

File: src/transformers/models/convnext/modeling_flax_convnext.py (hunk 8 above)
Comment (sanchit-gandhi) [reply src/transformers/models/convnext/modeling_flax_convnext.py:331]: Thanks for clarifying!

File: src/transformers/models/convnext/modeling_flax_convnext.py (hunk 4 above)
Comment (sanchit-gandhi) [reply src/transformers/models/convnext/modeling_flax_convnext.py]: Sorry we shouldn't remove `num_channels` from the config as this will break the PyTorch code!

File: src/transformers/models/convnext/configuration_convnext.py [hunk 14]
```diff
@@ -100,14 +99,12 @@ def __init__(
         layer_norm_eps=1e-12,
         layer_scale_init_value=1e-6,
         drop_path_rate=0.0,
-        image_size=224,
```
Comment (sanchit-gandhi) [inline src/transformers/models/convnext/configuration_convnext.py:103]: In general there shouldn't be a need to change the config file at all for a Flax port

File: src/transformers/models/convnext/configuration_convnext.py (hunk 13 above)
Comment (Shubhamai) [reply src/transformers/models/convnext/configuration_convnext.py:93]: Hi, I should have mentioned the PR wasn't fully complete, I am aware of the error due to failed tests, I will fix it asap ( probably on this weekend )

Comment (sanchit-gandhi) [reply src/transformers/models/convnext/configuration_convnext.py:93]: Cool! Sounds good! Feel free to ping me when it's ready and I'll get you a final review :)
